tianshou/policy/modelfree/c51.py

```diff
@@ -95,7 +95,7 @@ def learn(self, batch: Batch, **kwargs: Any) -> Dict[str, float]:
         cross_entropy = - (target_dist * torch.log(curr_dist + 1e-8)).sum(1)
         loss = (cross_entropy * weight).mean()
         # ref: https://github.com/Kaixhin/Rainbow/blob/master/agent.py L94-100
-        batch.weight = cross_entropy.detach()  # prio-buffer
+        batch.weight = cross_entropy.detach() * 1e9  # prio-buffer
```

Comment:
But I still don't quite get why it works. I know when using prio-buffer the current implementation will quickly reach loss≈0

Reply:
I suspect that other implementations have done some weight scaling on the fly. And I think I might have found the evidence:

Here the weights returned are normalized by max:
https://github.com/Kaixhin/Rainbow/blob/master/memory.py#L154
```python
weights = torch.tensor(weights / weights.max(), dtype=torch.float32, device=self.device)  # Normalise by max importance-sampling weight from batch
```

I'm running an experiment where I get ride of the `1e9` scaling on L98 but add a `weight = weight / weight.max()` in `prio.py`.

Reply:
Here are the results:

| Experiment         | Priority Buffer | NoisyLinear  | Dueling | Best reward |
| :---------------- | --------------:| -------------:| ------:| ------:|
| sanity check       | N | N | N | 16530 (C51 baseline: 16245) |
| prio only              | Y | N | N | 3982.5 |
| w/ weight scaling + beta=0.4 | Y | N | N | 16582.5 |
| weight scaling + beta=0.4 | Y | Y | Y |  11625  |
| w/ weight norm + beta=0.4 | Y | N | N | 17562.5 |
| weight norm + beta=0.4 | Y | Y | Y | 5035 |

It seems that weight scaling might be more consistent.

Reply:
> I suspect that other implementations have done some weight scaling on the fly.

Agree, but I personally do not favor the scaling fix (1e9). Adding such a magic number and this cannot find in all other implementations, that's pretty weird... Maybe we miss something?

I dig into stable-baselines implementation once again and they indeed have the eps-hack, but instead on DQN policy side: 
https://github.com/hill-a/stable-baselines/blob/f9be700e26a7eca9ec25e04003479910100d6662/stable_baselines/deepq/dqn.py#L299

Reply:
> > I suspect that other implementations have done some weight scaling on the fly.
> 
> Agree, but I personally do not favor the scaling fix (1e9). Adding such a magic number and this cannot find in all other implementations, that's pretty weird... Maybe we miss something?
> 
I agree scaling by a magic number is bad. I would rather do weight normalization.

> I dig into stable-baselines implementation once again and they indeed have the eps-hack, but instead on DQN policy side:
> https://github.com/hill-a/stable-baselines/blob/f9be700e26a7eca9ec25e04003479910100d6662/stable_baselines/deepq/dqn.py#L299
>
Great! So we'll definitely keep the eps hack. Having it on the buffer side is obviously better.

I tried `--training-num 4` and got some interesting results:

| Experiment         | Priority Buffer | NoisyLinear  | Dueling | Best reward |
| :---------------- | --------------:| -------------:| ------:| ------:|
| sanity check       | N | N | N | 16530 (C51 baseline: 16245) |
| prio only              | Y | N | N | 3982.5 |
| w/ weight scaling + beta=0.4 | Y | N | N | 16582.5 |
| weight scaling + beta=0.4 | Y | Y | Y |  11625  |
| w/ weight norm + beta=0.4 | Y | N | N | 17562.5 |
| weight norm + beta=0.4 | Y | Y | Y | 5035 |
| w/ weight norm + beta=0.4 + training-num=4 | Y | N | N | 17562.5 |
| weight norm + beta=0.4 + training-num=4 | Y | Y | Y | 14925 |

I'm a little surprised by the last result, which seems to suggest that weight normalization with proper parameters does work. Since we don't like scaling by a magic number, I think we should keep the weight normalization trick.

Reply:
> Here the weights returned are normalized by max:
> https://github.com/Kaixhin/Rainbow/blob/master/memory.py#L154

That's not returned weight, that's sampled weight

Reply:
I think it does. See https://github.com/Kaixhin/Rainbow/blob/9ff5567ad1234ae0ed30d8471e8f13ae07119395/agent.py#L63
```python
idxs, states, actions, returns, next_states, nonterminals, weights = mem.sample(self.batch_size)
```